feat: expose statuses to compile torch

fd0r opened the pull request.

RomanBredehoft commented on src/concrete/ml/quantization/quantized_module.py:
```diff
@@ -598,9 +599,13 @@ def compile(
                 error to a default value.
             verbose (bool): Indicate if compilation information should be printed
                 during compilation. Default to False.
+            statuses (Optional[Dict[str, str]]): same dict as Concrete Python encryption statuses
```
I would maybe better describe this, I think users should be able to understand what this is without having the need to look at CP !

RomanBredehoft replied:
like `Encryption statuses for all input values, either "clear" or "encrypted" `

RomanBredehoft commented on src/concrete/ml/torch/compile.py:
```diff
@@ -142,6 +143,7 @@ def _compile_torch_or_onnx_model(
         global_p_error (Optional[float]): probability of error of the full circuit. In FHE
             simulation `global_p_error` is set to 0
         verbose (bool): whether to show compilation information
+        statuses (Optional[Dict[str, str]]): same dict as Concrete Python encryption statuses
```
same

RomanBredehoft commented on src/concrete/ml/torch/compile.py:
```diff
@@ -229,6 +233,7 @@ def compile_torch_model(
         global_p_error (Optional[float]): probability of error of the full circuit. In FHE
             simulation `global_p_error` is set to 0
         verbose (bool): whether to show compilation information
+        statuses (Optional[Dict[str, str]]): same dict as Concrete Python encryption statuses
```
same

RomanBredehoft commented on tests/quantization/test_quantized_module.py:
```diff
@@ -351,6 +351,110 @@ def test_quantized_module_rounding_fhe(model_class, input_shape, default_configu
+    assert isinstance(numpy_input, numpy.ndarray)
+
+    # Compile with rounding activated
+    with pytest.raises(ValueError):
```
I would prefer `with pytest.raises(ValueError, match=...)` here, especially since we have two different types of value errors, it's more robust !

fd0r replied:
Indeed, I was a bit lazy there!
Will fix this

RomanBredehoft commented on tests/quantization/test_quantized_module.py:
```diff
@@ -351,6 +351,110 @@ def test_quantized_module_rounding_fhe(model_class, input_shape, default_configu
+        )
+
+    # No encrypted value
+    with pytest.raises(RuntimeError):
```
this is a CP restriction right ? is there a clear (🙂) message associated to it ? If not, maybe worth to add our own one, else maybe add a `match=` here as well

fd0r replied:
If there is no encrypted value CP can't compile indeed.

Their error message is quite explicit, I'll add a `match=""`

RomanBredehoft reviewed (changes requested):
thanks for this ! I have a few comments

jfrery commented on src/concrete/ml/quantization/quantized_module.py:
```diff
@@ -577,6 +577,7 @@ def compile(
         p_error: Optional[float] = None,
         global_p_error: Optional[float] = None,
         verbose: bool = False,
+        statuses: Optional[Dict[str, str]] = None,
```
I don't feel like statuses is very talkative. Should be referring to the input I think. `input_statuses` or `input_encryption` where the user passes booleans in a list following the order of the input in the forward call? so that you don't have to specify onnx node names.

jfrery reviewed (commented):
This could be needed indeed thanks!

fd0r commented:
Looks like KNN test is flaky

jfrery reviewed (approved):
Looks great thanks!

fd0r commented:
What I did:
- changed the name of the variable to a more explicit naming.
- updated the doc-strings to something clearer.
- added `match=` to the tests
- changed the type of the variable from a dict to a list to avoid having to handle onnx names

RomanBredehoft commented on src/concrete/ml/quantization/quantized_module.py:
```diff
@@ -598,9 +599,15 @@ def compile(
                 error to a default value.
             verbose (bool): Indicate if compilation information should be printed
                 during compilation. Default to False.
+            inputs_encryption_status (Optional[Sequence[str]]): encryption status ('clear',
```
I would still mention that everything is encrypted if the parameter is set to None !

fd0r replied:
It's mentionned in all the user facing APIs, I don't think many users will use this one tbh

RomanBredehoft replied:
I mean it was mostly for making the docstring consistent with what this parameters does, as we now provide a way to chose what inputs to encrypt, to avoid confusion that we don't chose which ones are encrypted or not based on whatever other rules. But sure if no one else pushes for this then go for it

RomanBredehoft commented on src/concrete/ml/torch/compile.py:
```diff
@@ -142,6 +143,8 @@ def _compile_torch_or_onnx_model(
         global_p_error (Optional[float]): probability of error of the full circuit. In FHE
             simulation `global_p_error` is set to 0
         verbose (bool): whether to show compilation information
+        inputs_encryption_status (Optional[Sequence[str]]): encryption status ('clear', 'encrypted')
```
same

fd0r replied:
Same what? There is a comment on the fact that by default everything is encrypted

RomanBredehoft replied:
same as in "same comment as above" in order to not avoid making changes in both places if there are some

RomanBredehoft commented on tests/quantization/test_quantized_module.py:
```diff
@@ -351,6 +355,103 @@ def test_quantized_module_rounding_fhe(model_class, input_shape, default_configu
     # TODO: https://github.com/zama-ai/concrete-ml-internal/issues/3800
 
 
+# TODO: extend this test with multi-input encryption status
```
worth a fimxe + link to an issue here right ?

RomanBredehoft reviewed (changes requested):
I still have a few comments

fd0r commented:
Adressed the feedback made by @RomanBredehoft

fd0r commented:
KNN flaky, relaunching

RomanBredehoft commented on src/concrete/ml/quantization/quantized_module.py:
```diff
@@ -598,9 +599,15 @@ def compile(
             Circuit: The compiled Circuit.
+
+        Raises:
+            ValueError: if inputs_encryption_status does not match with the
```
it's a bit more than that right ?

fd0r replied:
Well it's if the configuration is wrong, how would you formulate it?

RomanBredehoft replied:
I would usually very quickly describe what are bad configurations

RomanBredehoft reviewed (approved):
if no one pushes for these last comments, fine for me, thanks a lot for this feature !
